examples/rcn.py

```diff
@@ -0,0 +1,438 @@
+# ---
+# jupyter:
+#   jupytext:
+#     formats: ipynb,py:percent
+#     text_representation:
+#       extension: .py
+#       format_name: percent
+#       format_version: '1.3'
+#       jupytext_version: 1.13.2
+#   kernelspec:
+#     display_name: Python 3
+#     language: python
+#     name: python3
+# ---
+
+# %% [markdown]
+# # RCN implementation using the PGMax package
+
+# %% [markdown]
+# RCN is a neuroscience based, data-efficient, interpretable object detection/vision system. It is a probabilistic graphical model (PGM) that trains in minutes and generalizes well (for certain traits). It can reason about, and is robust to, occlusion. It can also imagine and interpret scenes.
+#
+# Please refer to the [Science](https://www.cs.jhu.edu/~ayuille/JHUcourses/ProbabilisticModelsOfVisualCognition2020/Lec21/A%20generative%20vision%20model%20that%20trains%20with%20high%20data%20efficiency%20and%20breaks%20text-based%20CAPTCHAs.pdf) paper for more details. In this notebook, we implement a two-level RCN model on a small subset of the MNIST dataset.
+
+# %%
+# %matplotlib inline
+import os
+import time
+
+import jax
+import matplotlib.pyplot as plt
+import numpy as np
+from jax import numpy as jnp
+from jax import tree_util
+from joblib import Memory
+
+memory = Memory("./example_data/tmp")
+
+from typing import Tuple
+
+from scipy.ndimage import maximum_filter
+from scipy.signal import fftconvolve
+from sklearn.datasets import fetch_openml
+
+from pgmax.fg import graph, groups
+
+fetch_openml_cached = memory.cache(fetch_openml)
+os.environ["XLA_PYTHON_CLIENT_ALLOCATOR"] = "platform"
+os.environ["XLA_PYTHON_CLIENT_PREALLOCATE"] = "false"
+
+# %% [markdown]
+# # 1. Load the data
+# %% [markdown]
+# In this notebook, we train an RCN model on a small subset of MNIST (training on 20 images and testing on 20 images). We also have support to train the model on 100 images. Use this option only if you have a gpu with atleast 8Gbs memory. Also, recommend that jax is installed with cuda enabled for training with more than 20 images.
+
+# %%
+train_size = 20
+test_size = 20
+
+
+# %%
+def fetch_mnist_dataset(test_size: int, seed: int = 5) -> Tuple[np.ndarray, np.ndarray]:
```

Comment:
Seems that using `tuple` was enabled in python 3.9?
https://www.python.org/dev/peps/pep-0585/

I was trying to use the suggestion but keep getting `TypeError: 'type' object is not subscriptable`. The project is using python version of `3.7.5`, maybe thats why?

Reply:
Ahh - my bad. You need to add:
```
from __future__ import annotations
```
to the top of the file - it'll be default in future python version
You often want that for other reasons, as various other typing things will fail otherwise.

Reply:
Fixed :D